### PR TITLE
MB-59616: Adding vector_base64 field

### PR DESCRIPTION
**Pull request by Likith101**
- Added a new field type called vector_base64.
 - Acts similar to vector in most cases.
 - When a new document arrives in the bleve layer, during the parsing of all its fields in processProperty, if the field mapping type is vector-base64, then its value is decoded into a vector field and processed like a vector.
 - The standard golang base64 library is used for the decode operation.

### CODE REVIEW

File: document/field_vector_base64.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+func (n *VectorBase64Field) NumPlainTextBytes() uint64 {
+	return n.vectorField.NumPlainTextBytes()
+}
+
+func (n *VectorBase64Field) AnalyzedLength() int {
+	return n.vectorField.AnalyzedLength()
+}
+
+func (n *VectorBase64Field) EncodedFieldType() byte {
+	return 'e' // CHECK
```
Comment (abhinavdangeti) [inline document/field_vector_base64.go:58]: Why `e`?

File: document/field_vector_base64.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+func (n *VectorBase64Field) AnalyzedTokenFrequencies() index.TokenFrequencies {
+	return n.vectorField.AnalyzedTokenFrequencies()
+}
+
+func (n *VectorBase64Field) Analyze() {
+	// CHECK
```
Comment (abhinavdangeti) [inline document/field_vector_base64.go:66]: Drop this comment, no analysis needed here.

File: document/field_vector_base64.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+
+	index "github.com/blevesearch/bleve_index_api"
+)
+
+type VectorBase64Field struct {
+	vectorField  *VectorField
+	encodedValue string
```
Comment (abhinavdangeti) [inline document/field_vector_base64.go:30]: Rename to base64Encoding

File: document/field_vector_base64.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+func (n *VectorBase64Field) GoString() string {
+	return fmt.Sprintf("&document.vectorFieldBase64Field{Name:%s, Options: %s, "+
+		"Value: %+v}", n.vectorField.Name(), n.vectorField.Options(), n.vectorField.Value())
+}
+
+// For the sake of not polluting the API, we are keeping arrayPositions as a
+// parameter, but it is not used.
+func NewVectorBase64Field(name string, arrayPositions []uint64, encodedValue string,
```
Comment (abhinavdangeti) [inline document/field_vector_base64.go:80]: encodedValue -> vectorBase64

File: mapping/mapping_vectors.go [hunk 5]
```diff
@@ -121,6 +134,27 @@ func processVector(vecI interface{}, dims int) ([]float32, bool) {
 	return rv, true
 }
 
+func processVectorBase64(vecBase64 interface{}) (interface{}, bool) {
+
```
Comment (abhinavdangeti) [inline mapping/mapping_vectors.go:138]: Drop empty line.

File: mapping/mapping_vectors.go [hunk 6]
```diff
@@ -140,13 +174,24 @@ func (fm *FieldMapping) processVector(propertyMightBeVector interface{},
 	return true
 }
 
+func (fm *FieldMapping) processVectorBase64(propertyMightBeVectorBase64 interface{},
+	pathString string, path []string, indexes []uint64, context *walkContext) {
+
```
Comment (abhinavdangeti) [inline mapping/mapping_vectors.go:179]: Drop empty line.

File: search_knn.go [hunk 7]
```diff
@@ -67,10 +68,11 @@ type SearchRequest struct {
 }
 
 type KNNRequest struct {
-	Field  string       `json:"field"`
-	Vector []float32    `json:"vector"`
-	K      int64        `json:"k"`
-	Boost  *query.Boost `json:"boost,omitempty"`
+	Field        string       `json:"field"`
+	Vector       []float32    `json:"vector"`
+	VectorBase64 string       `json:"vectorbase64"`
```
Comment (abhinavdangeti) [inline search_knn.go:73]: `json:"vector_base64"`

File: search_knn_test.go [hunk 8]
```diff
@@ -397,6 +398,162 @@ func min(a, b int) int {
 	return b
 }
 
+func TestVectorBase64Index(t *testing.T) {
+
```
Comment (abhinavdangeti) [inline search_knn_test.go:402]: Drop empty line

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: document/field_vector_base64.go (hunk 1 above)
Comment (Likith101) [reply document/field_vector_base64.go:58]: Went with e for encoded. B is taken for boolean. Or should I use the same type as vector with is v?

File: mapping/mapping_vectors.go (hunk 6 above)
Comment (abhinavdangeti) [reply mapping/mapping_vectors.go:178]: Missed this.

File: document/field_vector_base64.go (hunk 1 above)
Comment (abhinavdangeti) [reply document/field_vector_base64.go]: Hmm ok, I don't have a strong opinion on this - do anyone else do?

Comment (abhinavdangeti) [reply document/field_vector_base64.go]: Shouldn't use `v` though - that's taken.

Comment (abhinavdangeti) [reply document/field_vector_base64.go]: Drop the `CHECK` here.

File: document/field_vector_base64.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	return &VectorBase64Field{
+		vectorField: NewVectorFieldWithIndexingOptions(name, arrayPositions,
+			vector, dims, similarity,
+			vectorIndexOptimizedFor, DefaultVectorIndexingOptions),
+
+		encodedValue: encodedValue,
+	}, nil
+}
+
+func DecodeVector(encodedValue string) ([]float32, error) {
```
Comment (abhinavdangeti) [inline document/field_vector_base64.go:97]: Let's add some commentary around this algorithm here.

File: mapping/mapping_vectors.go [hunk 10]
```diff
@@ -121,6 +134,27 @@ func processVector(vecI interface{}, dims int) ([]float32, bool) {
 	return rv, true
 }
 
+func processVectorBase64(vecBase64 interface{}) (interface{}, bool) {
```
Comment (abhinavdangeti) [inline mapping/mapping_vectors.go:137]: The code within here looks exactly as in `document.DecodeVector(..)`. Let's just reuse that here after you do the type assertion below - where you can interpret an `err != nil` as false.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: mapping/mapping_vectors.go [hunk 11]
```diff
@@ -175,11 +153,17 @@ func (fm *FieldMapping) processVector(propertyMightBeVector interface{},
 
 func (fm *FieldMapping) processVectorBase64(propertyMightBeVectorBase64 interface{},
 	pathString string, path []string, indexes []uint64, context *walkContext) {
-	propertyMightBeVector, ok := processVectorBase64(propertyMightBeVectorBase64)
+
```
Comment (abhinavdangeti) [inline mapping/mapping_vectors.go:156]: That empty line popped up again.

File: document/field_vector_base64.go [hunk 12]
```diff
@@ -94,16 +94,26 @@ func NewVectorBase64Field(name string, arrayPositions []uint64, vectorBase64 str
...
+	// We first decode the encoded string into a byte array.
 	decodedString, err := base64.StdEncoding.DecodeString(encodedValue)
 	if err != nil {
-		fmt.Println("Error decoding string:", err)
 		return nil, err
 	}
 
+	// The array is expected to be divisible by 4 because each float32
+	// should occupy 4 bytes
+	if len(decodedString)%4 != 0 {
```
Comment (abhinavdangeti) [inline document/field_vector_base64.go:109]: Like we decided, let's use `size.SizeOfFloat32` here in place of 4.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]
@Likith101 looking mostly ready. Would you add a benchmark test to here that tests time taken for `DecodeVector(..)` for a few dimensionalities - maybe 128, 784, 1536.

File: document/field_vector_base64.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// The array is expected to be divisible by 4 because each float32
+	// should occupy 4 bytes
+	if len(decodedString)%4 != 0 {
+		return nil, fmt.Errorf("Decoded byte array not divisible by 4")
+	}
+	dims := int(len(decodedString) / 4)
```
Comment (Thejas-bhat) [inline document/field_vector_base64.go:112]: does this need to be `len(decodedString) / size.SizeOfFloat32` as well?

**Comment by abhinavdangeti** [comment]
@Likith101 faiss_vector_base64_test.go also should the `//go:build vectors` treatment :)

**Comment by abhinavdangeti** [comment]
@Likith101 Did you already run the decode algorithm by Jon?

**Comment by Likith101** [comment]
> @Likith101 Did you already run the decode algorithm by Jon? Would you also update the description above the benchmark test results?

I ran the decode algorithm by Jon, so no changes necessary there.

**Comment by Likith101** [comment]
These are the benchmark results for the decode function done on my local system.

| Dimensions | Num Runs Per Sec | Num Secs Per Op (ns) | 
| --- | --- | --- |
| 128 | 2184279 | 541.8 |
| 784 | 410046 | 2964 |
| 1536 | 212860 | 5569 |

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**